File: 2025_06/leetcode2025-06-20.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int maxDistance(string s, int k) {
        int n = s.size();
        unordered_map<char, int> cnt;
        int res = 0;
        for (int i=0; i<n; i++){
            cnt[s[i]]++;
            int x = abs(cnt['E']-cnt['W']);
            int y = abs(cnt['N']-cnt['S']);
            int canAdd = 2*min(k, min(cnt['E'], cnt['W'])+min(cnt['N'], cnt['S']));
            res = max(res, x+y+canAdd);
        }
        return res;
    }
};

int main(){
    Solution sl;
    string s = "NWSE";
    int k = 1;
    cout << sl.maxDistance(s,k) << endl;
}

File: leetcode2025/2025_06/leetcode2025-06-20.py
from collections import defaultdict

class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        dic = defaultdict(int)
        res = 0
        for i in s:
            dic[i] += 1
            y = abs(dic["N"]-dic["S"])
            x = abs(dic["E"]-dic["W"])
            tmp = x+y
            tmp += 2*min(k,min(dic["E"],dic["W"])+min(dic["N"],dic["S"]))
            res = max(res, tmp)
        return res
            

if __name__ == "__main__":
    sl = Solution()
    s = "NWSE"
    k = 1
    print(sl.maxDistance(s, k))
